### PR TITLE
Incremental backup: fix calculation of binlog files to use

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Incremental backup logic introduced in https://github.com/vitessio/vitess/pull/11097 (still undocumented) did not evaluate the correct list of binary logs when `gtid_purged` happened to be exactly the last full backup position. There are probably more scenarios affected by the bug.

This PR treats `gtid_purged` and the last known backup pos correctly. Tests added.

Incremental backup and Point in time recovery as introduced in https://github.com/vitessio/vitess/pull/11097, are still undocumented.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/13060
- Related to: https://github.com/vitessio/vitess/pull/11097

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 1]
```diff
@@ -71,43 +71,42 @@ func ChooseBinlogsForIncrementalBackup(
...
+			// All previous binary logs are fully contained by backupPos. So definitely all binlogs _prior_ to
+			// this binlog are not necessary. We still don't know about _this_ binlog. We can't tell yet if
+			// _this_ binlog introduces new GTID entries not covered by the last backup pos. But we will only
+			// know this when we look into the _next_ binlog file's Previous-GTIDs.
+			continue
 		}
-		if unionPreviousGTIDs {
-			prevPos.GTIDSet = prevGTIDsUnion
+		if i == 0 {
+			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "the very first binlog file %v has PreviousGTIDs %s that exceed given incremental backup pos. There are GTID entries that are missing and this backup cannot run", binlog, previousGTIDsPos)
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid.go:98]: Might be worth using the GTID_PURGED language/variable specifically to aid in debugging and bridging the MySQL<->Vitess concepts/pieces here.

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 2]
```diff
@@ -71,43 +71,42 @@ func ChooseBinlogsForIncrementalBackup(
...
 		}
-		if !prevPos.GTIDSet.Contains(lookFromGTIDSet) {
-			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "binary log %v with previous GTIDS %s neither contains requested GTID %s nor contains it. Backup cannot take place", binlog, prevPos.GTIDSet, lookFromGTIDSet)
+		if !prevGTIDsUnion.Union(purgedGTIDSet).Contains(backupFromGTIDSet) {
+			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "binary log %v with previous GTIDS %s neither contains requested GTID %s nor contains it. Backup cannot take place", binlog, previousGTIDsPos.GTIDSet, backupFromGTIDSet)
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid.go:101]: I'm not sure what `neither contains requested GTID %s nor contains it` means here.

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 3]
```diff
@@ -71,43 +71,42 @@ func ChooseBinlogsForIncrementalBackup(
...
+		// The binary logs are read in-order. They expand. For example, we know
+		// Previous-GTIDs of binlog file 0000018 contain those of binlog file 0000017.
+		// We look for the first binary log whose Previous-GTIDs isn't already fully covered
+		// by "backupPos" (the position from which we want to create the incremental backup).
 		// That means the *previous* binary log is the first binary log to introduce GTID events on top
 		// of "backupPos"
-		if i == 0 {
-			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "the very first binlog file %v has PreviousGTIDs %s that exceed given incremental backup pos. There are GTID entries that are missing and this backup cannot run", binlog, prevPos)
+		if backupFromGTIDSet.Contains(previousGTIDsPos.GTIDSet) {
+			// All previous binary logs are fully contained by backupPos. So definitely all binlogs _prior_ to
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid.go:91]: IMO it's easier to reason about when explained as the current binary log being the oldest one available -- so all previous GTIDs have been purged and are no longer available.

File: go/vt/mysqlctl/binlogs_gtid_test.go [hunk 4]
```diff
@@ -109,15 +120,81 @@ func TestChooseBinlogsForIncrementalBackup(t *testing.T) {
...
+				"vt-bin.000001": "16b1039f-22b6-11ed-b765-0a43f95f28a3:3-78",
+				"vt-bin.000002": "16b1039f-22b6-11ed-b765-0a43f95f28a3:3-90",
+				"vt-bin.000003": "16b1039f-22b6-11ed-b765-0a43f95f28a3:3-90",
+				"vt-bin.000004": "16b1039f-22b6-11ed-b765-0a43f95f28a3:3-100",
+				"vt-bin.000005": "16b1039f-22b6-11ed-b765-0a43f95f28a3:3-110",
+				"vt-bin.000006": "16b1039f-22b6-11ed-b765-0a43f95f28a3:2-300",
+			},
+			backupPos:   "16b1039f-22b6-11ed-b765-0a43f95f28a3:1-63",
+			gtidPurged:  "16b1039f-22b6-11ed-b765-0a43f95f28a3:1-2",
+			expectError: "There are GTID entries that are missing",
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid_test.go:156]: Nit, but `[purged and] no longer available` feels more accurate than missing.

**Review by mattlord — APPROVED** [review]
LGTM! Only had a few very minor nits/suggestions. I'll leave those up to you.

File: go/vt/mysqlctl/binlogs_gtid.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/binlogs_gtid.go]: Reworded

**Comment by shlomi-noach** [comment]
@mattlord please see if rewording makes more sense. I find it really difficult to explain the constraints in proper English. Please feel free to edit.

**Comment by mattlord** [comment]
> @mattlord please see if rewording makes more sense. I find it really difficult to explain the constraints in proper English. Please feel free to edit.

@shlomi-noach LGTM! ❤️

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 5]
```diff
@@ -71,43 +71,44 @@ func ChooseBinlogsForIncrementalBackup(
...
+			// Previous-GTIDs is contained by backupPos. So definitely all binlogs _prior_ to
+			// this binlog are not necessary. We still don't know about _this_ binlog. We can't tell yet if
+			// _this_ binlog introduces new GTID entries not covered by the last backup pos. But we will only
+			// know this when we look into the _next_ binlog file's Previous-GTIDs.
+			continue
 		}
-		if unionPreviousGTIDs {
-			prevPos.GTIDSet = prevGTIDsUnion
+		if i == 0 {
+			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "Required entries have been purged. Oldest binary log %v expects entries not found in back up pos. Expected pos=%v", binlog, previousGTIDsPos)
```
Comment (deepthi) [inline go/vt/mysqlctl/binlogs_gtid.go:98]:
```suggestion
			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "Required entries have been purged. Oldest binary log %v expects entries not found in backup pos. Expected pos=%v", binlog, previousGTIDsPos)
```

Comment (deepthi) [reply go/vt/mysqlctl/binlogs_gtid.go:98]: I wonder if `pos` in the error message has a clear meaning for users. It seems like terminology that only people familiar with code will understand. What is a good alternative?

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 6]
```diff
@@ -71,43 +71,44 @@ func ChooseBinlogsForIncrementalBackup(
...
 		}
-		if !prevPos.GTIDSet.Contains(lookFromGTIDSet) {
-			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "binary log %v with previous GTIDS %s neither contains requested GTID %s nor contains it. Backup cannot take place", binlog, prevPos.GTIDSet, lookFromGTIDSet)
+		if !prevGTIDsUnion.Union(purgedGTIDSet).Contains(backupFromGTIDSet) {
+			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION,
+				"Mismatching GTID entries. Requested backup pos has entries not found in the binary logs, and binary logs have entries not found in the requested backup pos. Requested pos=%v, binlog pos=%v",
```
Comment (deepthi) [inline go/vt/mysqlctl/binlogs_gtid.go:102]: Is it `and binary logs` or `or binary logs`?
In other words, does the check imply both conditions have occurred, or that at least one of them has occurred?
You could replace `and` with `and/or`.

**Review by deepthi — APPROVED** [review]
Only nits on wording etc. The logic looks good to me.

File: go/vt/mysqlctl/binlogs_gtid.go (hunk 6 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/binlogs_gtid.go:102]: It's an `and`, as described. It's like a Venn diagram of two circles which overlap, but neither fully contains the other.

Comment (shlomi-noach) [reply go/vt/mysqlctl/binlogs_gtid.go:102]: Added `Neither fully contains the other` to the message. I hope it clarifies.

**Pull request merged**